vmm/sandbox/src/vm.rs

```diff
@@ -72,6 +72,26 @@ pub trait VM: Serialize + Sync + Send {
     async fn wait_channel(&self) -> Option<Receiver<(u32, i128)>>;
 }
 
+#[macro_export]
+macro_rules! impl_recoverable {
+    ($ty:ty) => {
+        #[async_trait]
+        impl $crate::vm::Recoverable for $ty {
+            async fn recover(&mut self) -> Result<()> {
+                self.client = Some(self.create_client().await?);
+                let pid = self.pid().await?;
+                let (tx, rx) = channel((0u32, 0i128));
+                tokio::spawn(async move {
+                    let wait_result = wait_pid(pid as i32).await;
+                    tx.send(wait_result).unwrap_or_default();
+                });
+                self.wait_chan = Some(rx);
+                Ok(())
+            }
+        }
+    };
+}
```

Comment:
Based on the comments above, I don't think the recovering process must be very different from each other, if it is so, maybe extracting this macro is not necessary. 

If the recovering is so common, that the recover codes is duplicated, maybe there is another way to do it.
```
impl<T: VM> Recoverable for T {
    async fn recover(&mut self) -> Result<()>
}

```

and add `create_client` and `pid` to the trait of VM, so that the compiler could easily checkout the constraint that we can only recover the object with `create_client` and `pid` method.